### PR TITLE
Deployment profile testing with BATS

### PR DESCRIPTION
**Pull request by jandubois**
This PR contains both the BATS framework for testing deployment profiles, and a sample test to demonstrate how to use it.

It works for macOS using user profiles. Testing system profiles will require password-less sudo.

The code has not been tested on Windows yet. It is only expected to work for the user-legacy location on Windows; every other location requires admin access and is out of scope for this PR.

### CODE REVIEW

File: bats/tests/helpers/profile.bash [hunk 1]
```diff
@@ -0,0 +1,284 @@
+if is_macos; then
+    PROFILE_SYSTEM_DEFAULTS=/Library/Preferences/io.rancherdesktop.profile.defaults.plist
+    PROFILE_SYSTEM_LOCKED=/Library/Preferences/io.rancherdesktop.profile.locked.plist
+    PROFILE_USER_DEFAULTS="${HOME}${PROFILE_SYSTEM_DEFAULTS}"
+    PROFILE_USER_LOCKED="${HOME}${PROFILE_SYSTEM_LOCKED}"
+fi
+
+if is_linux; then
+    PROFILE_SYSTEM_DEFAULTS=/etc/rancher-desktop/defaults.json
+    PROFILE_SYSTEM_DEFAULTS=/etc/rancher-desktop/locked.json
```
Comment (IsaSih) [inline bats/tests/helpers/profile.bash:10]: variable name should be PROFILE_SYSTEM_LOCKED

File: bats/tests/helpers/profile.bash [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+export_profile() {
+    local dir=$1
+    if profile_exists; then
+        local dest="${dir}/profile.${PROFILE_LOCATION}.${PROFILE_TYPE}"
+        local profile=$(PROFILE_PATH="" profile_location)
+        if is_windows; then
+            reg.exe export "$profile" "${dest}.reg" /y
+        else
+            # Keep .plist or .json file extension
+            cp "$profile" "${dest}.${profile##*.}"
```
Comment (IsaSih) [inline bats/tests/helpers/profile.bash:103]: Please, add a -rf so that errors in exporting profile do not fail the test

**Review by IsaSih — CHANGES_REQUESTED** [review]

File: bats/tests/helpers/vm.bash [hunk 3]
```diff
@@ -60,9 +60,11 @@ start_container_engine() {
     local args=(
         --application.debug
         --application.updater.enabled=false
-        --container-engine.name="$RD_CONTAINER_ENGINE"
         --kubernetes.enabled=false
     )
+    if [ -n "$RD_CONTAINER_ENGINE" ]; then
```
Comment (IsaSih) [inline bats/tests/helpers/vm.bash:65]: why did you move $RD_CONTAINER_ENGINE away from the list of args? it will always be set, even if not specified, the default value applies, and it's required to avoid the first-run dialog

File: bats/tests/helpers/defaults.bash [hunk 4]
```diff
@@ -45,13 +45,28 @@ using_ghcr_images() {
...
 }
 
+########################################################################
+# RD_USE_PROFILE is for internal use. It uses a profile instead of
+# settings.json to set initial values for WSL integrations and allowed
+# images list because when settings.json exists the default profile is
+# ignored.
+
+: "${RD_USE_PROFILE:=false}"
+
```
Comment (IsaSih) [inline bats/tests/helpers/defaults.bash:69]: suggest to create function 
```
using_profiles() {
is_true "$RD_USE_PROFILE"}
```

File: bats/tests/helpers/vm.bash [hunk 5]
```diff
@@ -92,16 +94,25 @@ start_container_engine() {
...
     registry="docker.io"
     if using_ghcr_images; then
         registry="ghcr.io"
     fi
-    if is_windows; then
-        wsl_integrations="{\"$WSL_DISTRO_NAME\":true}"
-    fi
-    mkdir -p "$PATH_CONFIG"
-    cat <<EOF >"$PATH_CONFIG_FILE"
+    if is_true "${RD_USE_PROFILE-}"; then
```
Comment (IsaSih) [inline bats/tests/helpers/vm.bash:101]: if `using_profile` , as suggested to be created in line 69

**Review by IsaSih — CHANGES_REQUESTED** [review]

File: bats/tests/helpers/profile.bash [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+# Execute command for each profile
+foreach_profile() {
+    local locations=("$PROFILE_SYSTEM" "$PROFILE_USER")
+    if is_windows; then
+        locations+=("$PROFILE_SYSTEM_LEGACY" "$PROFILE_USER_LEGACY")
+    fi
+
+    local PROFILE_LOCATION PROFILE_TYPE
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash:61]: Because these are the names we ultimately need to set before invoking the command. I originally passed them like this:

```
PROFILE_LOCATION=$location PROFILE_TYPE=$type "$@"
```

because I only want to set them for the call, but restore them to their original values when `foreach_profile` returns. Then I realized that `local` would do that for me automatically, so I can just set them here without the additional variables.

File: bats/tests/helpers/profile.bash [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+# Check if profile exists
+profile_exists() {
+    case $OS in
+    darwin | linux)
+        [[ -f $(profile_location) ]]
+        ;;
+    windows)
+        profile_reg query &>/dev/null
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash:76]: No, `profile_exists` is a predicate function, it is supposed to return true/false and is used like this:

```shell
if profile_exists; then
    ...
fi
```

If you want to use it with `assert`, then you have to use it like this:

```shell
run profile_exists
assert_success
```

File: bats/tests/helpers/profile.bash [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+profile_jq() {
+    local expr=$1
+    local filename=$(profile_location)
+    # Need to use a temp file to avoid truncating the file before it has been read.
+    jq "$expr" "$filename" | profile_cat "${filename}.tmp"
+    profile_sudo mv "${filename}.tmp" "$filename"
+}
+
+# Usage: profile_plutil $action $options
+#
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash:260]: Well, the difference is that in this case the `$setting` are part of the `$options` that are being passed on to `plutil`. We only inspect the value to create the additional dictionaries.

For `profile_reg add|delete` the `$setting` parameter is extra, and we remove it from the options before passing them on.

So I think this is different and shouldn't be called out explicitly.

File: bats/tests/helpers/profile.bash [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    # Make sure all the dictionaries for the setting path exist
+    if [[ $action =~ ^-insert|-replace|-remove$ ]]; then
+        local setting=$1
+        local count=$(count_setting_segments "$setting")
+        if [[ $count -gt 1 ]]; then
+            local index
+            for index in $(seq $((count - 1))); do
+                local keypath=$(echo "$setting" | cut -d . -f 1-"$index")
+                # Ignore error if dictionary already exists
+                run profile_sudo plutil -insert "$keypath" -dictionary
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash]: Yes indeed, it does. Good catch!

I don't know how the sample test used to work before. I guess I must have made changes without running the test again.

Fixed.

File: bats/tests/helpers/profile.bash [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+            done
+        fi
+    fi
+
+    profile_sudo plutil "$action" "$@" "$(profile_location)"
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash]: Because we call `shift` at the top, removing `$1` from `$@`. I've changed it so we no longer do that.

File: bats/tests/helpers/profile.bash [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+
+        local reg_value_name=$(echo "$setting" | cut -d . -f "$count")
+        # reg_value_name may be empty when deleting a registry key instead of a named value
+        if [[ -n $reg_value_name ]]; then
+            # turn protected dots back into regular dots again
+            set - /v "${reg_value_name//$RD_PROTECTED_DOT/.}" "$@"
+        fi
+
+        # Overwrite existing entries without prompt
+        if [[ $action == add ]]; then
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/profile.bash]: Currently we pass in the `/f` for all delete actions, so it makes sense to just move it into the wrapper and remove it from the call sites.

**Review by jandubois — COMMENTED** [review]
Updated the PR with some of the suggested changes; see replies for the rest.

I also had to update the `start_app` function in the sample test; not sure why this stopped working. The change is similar to something I did in #5108.

We can later combine it, but once #5104 is merged this may be further simplified again.

**Review by IsaSih — CHANGES_REQUESTED** [review]

File: bats/tests/helpers/utils.bash [hunk 12]
```diff
@@ -39,6 +47,57 @@ assert_nothing() {
...
+# Convert raw string into properly quoted JSON string
+json_string() {
+    echo -n "$1" | jq -Rr @json
+}
+
+# Join list elements by separator after converting them via the mapping function
+# Examples:
+#   join "/" echo usr local bin            =>   usr/local/bin
+#   join ", " json_string a b\ c\"d\\e f   =>   "a", "b c\"d\\e", "f"
+join() {
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/utils.bash]: It is now called `join_map`

File: bats/tests/helpers/utils.bash [hunk 13]
```diff
@@ -39,6 +47,57 @@ assert_nothing() {
...
+    local sep=$1
+    local map=$2
+    shift 2
+
+    local elem result
+    for elem in "$@"; do
+        elem=$(eval "$(printf "%s %q" "$map" "$elem")")
```
[the remark this answers is not in the document]
Comment (jandubois) [reply bats/tests/helpers/utils.bash]: Indeed. Replaced by simpler expression

**Review by IsaSih — APPROVED** [review]
